### PR TITLE
Adds messages and chat command

### PR DESCRIPTION
**Pull request by sanderpick**
- `messages` is like `comments` without a target, added to top level of thread
- leverages `messages` and `sub` commands into `chat` 💯

### CODE REVIEW

**Comment by sanderpick** [comment]
Couldn't resist. Took a break from deploying. Will pick it up tomorrow.

**Comment by sanderpick** [comment]
Check it out... threads chat, recovering from a member going offline. This feels like the culmination of the threads and cafe protocols.

screenshot 2018-12-04 20 51 14

File: cmd/sub.go [hunk 1]
```diff
@@ -16,68 +15,111 @@ func init() {
 type subCmd struct {
 	Client ClientOptions `group:"Client Options"`
 	Thread string        `short:"t" long:"thread" description:"Thread ID. Omit for all."`
-	Type   string        `short:"k" long:"type" description:"Comma-separated list of event types to include. Omit for all."`
+	Type   []string      `short:"k" long:"type" description:"An update type to filter for. Omit for all."`
```
Comment (carsonfarmer) [inline cmd/sub.go]: doh, didn't realize this was possible. this is what i wanted all along! nice 👍

File: cmd/sub.go [hunk 2]
```diff
@@ -16,68 +15,111 @@ func init() {
...
+			output(string(data))
+		}
 	}
-	defer req.Body.Close()
-	if req.StatusCode >= 400 {
-		res, err := unmarshalString(req.Body)
+	return nil
+}
+
+func callSub(threadId string, types []string) (<-chan core.ThreadUpdate, error) {
```
Comment (carsonfarmer) [inline cmd/sub.go]: yeh this is nice, glad its proving useful outside of demos!

File: cmd/sub.go [hunk 3]
```diff
@@ -16,68 +15,111 @@ func init() {
...
-	}
-	decoder := json.NewDecoder(req.Body)
-	for {
-		var info core.ThreadUpdate
-		if err := decoder.Decode(&info); err == io.EOF {
-			break
-		} else if err != nil {
-			return err
+
+		if req.StatusCode >= 400 {
```
Comment (carsonfarmer) [inline cmd/sub.go]: ah good catch, this came up today during some tests actually 👍

File: core/api.go [hunk 4]
```diff
@@ -132,6 +133,12 @@ func (a *api) Start() {
 			}
 		}
 
+		messages := v0.Group("/messages")
```
Comment (carsonfarmer) [inline core/api.go]: this feels a bit weird having the two separate paths like this? not sure how else it should be done... is this pretty standard practice?

File: core/api_messages.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
+package core
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/gin-gonic/gin"
+)
+
+func (a *api) addThreadMessages(g *gin.Context) {
```
Comment (carsonfarmer) [inline core/api_messages.go]: so good!

File: cmd/chat.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+func (x *chatCmd) Long() string {
+	return `
+Starts an interactive chat session in a thread.
+Omit the --thread option to use the default thread (if selected).
+`
+}
+
+func (x *chatCmd) Execute(args []string) error {
```
Comment (carsonfarmer) [inline cmd/chat.go]: siiiick!

File: core/api.go (hunk 4 above)
Comment (sanderpick) [reply core/api.go]: yeah, agreed. same for files. my thought there was it's nice to be able to paginate across all files / messages... but maybe not needed. can revisit.

**Comment by sanderpick** [comment]
Cheers for the look @carsonfarmer !

**Pull request merged**